nixos/qemu-vm: Option to use squashfs Nix store closure instead of virtfs access to host's store

chkno opened the pull request:
###### Motivation for this change
Allow VMs built with `nixos-rebuild build-vm` to work without any virtfs mounts into the host.  In particular, allow them to use a nix store closure in a squashfs like the installer uses.

This allows better isolation of guests.  A host that runs multiple VMs may not wish the guests to be able to see each others' derivations.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review rev --branch staging HEAD"` -- It's zero packages, but this refactoring touches all the NixOS tests.  I ran a handful of existing ones.
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after) -- N/A?
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

chkno committed:
nixos/qemu-vm: Test for shareNixStore = false

chkno commented:
(Rebased on top of #72342 so the NixOS tests can run)

chkno commented:
@GrahamcOfBorg test qemu-private-store

matthewbauer commented:
@grahamcofborg test qemu-private-store

matthewbauer reviewed (commented):
This all looks like a good idea, though I don't have enough familiarity to really judge whether this breaks anything. Requested a few more people to review.

chkno committed:
nixos/qemu-vm: Option to not share Nix store with host

When the store is not shared, the closure used in the VM it is packaged
up in a squashfs, just like the ISO installer does.

chkno committed:
nixos/qemu-vm: New options for turning virtfs mounts off

chkno commented:
Updated to rebase on top of #55645.  I confirm that login, upnp, installer, and the new test pass on x86_64-linux.

matthewbauer reviewed (approved):
Looks good!

chkno commented:
Updated test from perl test framework to python test framework.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review-may-2019/3032/157

flokli commented on nixos/modules/virtualisation/qemu-vm.nix:
```diff
@@ -282,16 +308,50 @@ in
           '';
       };
 
+    virtualisation.shareExchangeDir =
+      mkOption {
+        default = true;
+        type = types.bool;
+        description =
+          ''
+            Share /tmp/xchg with the host.
+          '';
+      };
+
+    virtualisation.shareSharedDir =
+      mkOption {
+        default = true;
+        type = types.bool;
+        description =
+          ''
+            Share /tmp/shared with the host.
+          '';
+      };
```
What's the difference of these?

chkno replied:
I don't know.

By default they point at the same directory in the host.  The test framework points them at different directories by setting `SHARED_DIR`.

History:
* Aug 2011 d75efe4aa184c4fa89241cf3df2f04616f3ba93f edolstra: Introduced /tmp/xchg: "For security, don't mount the entire host filesystem."
* Aug 2012 a025e7e7e287c2539d883bd40a769ae29f9087ff edolstra: Introduced /tmp/shared: "Provide a common share between VMs to allow easy communication"
* May 2020 d85dc4f6902819be8277d1d65414e51bfbef0ac6 erikarvstedt: Mentions "xchg is advertised as a bidirectional exchange dir."

Current uses I could find:
| Mount | Purpose | Location |
| --- | --- | --- |
| xchg | coverage-data | nixos/modules/testing/test-instrumentation.nix nixos/lib/test-driver/test-driver.pl |
| xchg | runInMachine | nixos/lib/testing.nix |
| shared | copy_from_host copy_from_vm | nixos/lib/test-driver/test-driver.py |


I just want a lever to turn them off.  I guess it makes more sense to be one lever rather than two.

I combined these two options into one and added a note about what these directories are currently used for in the option description.

flokli replied:
@erikarvstedt @edolstra can you provide some insight here?

edolstra replied:
We can probably get rid of `coverage-data`. AFAIK we're not doing coverage runs in VMs anymore (and IIRC it was only used in some toy examples like the Quake 3 test anyway).

edolstra replied:
We can probably also eliminate `copy_from_*` since it's only used in 3 places, which can be replaced by `machine.execute("cat ...")`.

flokli replied:
> We can probably get rid of `coverage-data`. AFAIK we're not doing coverage runs in VMs anymore (and IIRC it was only used in some toy examples like the Quake 3 test anyway).

I updated the list in https://github.com/NixOS/nixpkgs/issues/72828, and filed https://github.com/NixOS/nixpkgs/pull/97231 removing the remaining `coverage-data` references - PTAL.

flokli commented:
This also needs a rebase.

chkno commented:
Rebased on top of #92205, #92122, and #90280.

nh2 commented on nixos/modules/virtualisation/qemu-vm.nix:
```diff
@@ -348,16 +352,41 @@ in
+        type = types.bool;
+        description =
+          ''
+            Share /tmp/shared and /tmp/xchg with the host.  These are used
+            for runInMachine, copy_from_{host,vm}, and coverage-data.
```
May want to use `<literal>` markup for easier reading, e.g.

```suggestion
            Share <literal>/tmp/shared</literal> and <literal>/tmp/xchg</literal> with the host.
            These are used for <literal>runInMachine</literal>, <literal>copy_from_{host,vm}</literal>,
            and <literal>coverage-data</literal>.
```

(Please check the manual builds after this, one can accidentlaly screw it up like I did [recently](https://github.com/NixOS/nixpkgs/pull/92240#issuecomment-653740926).)

aszlig replied:
@nh2: ... or go even further and use `<filename/>` and `<function/>`:

```suggestion
            Share <filename class="directory">/tmp/shared</filename> and
            <filename class="directory">/tmp/xchg</filename> with the host.
            These are used for <function>runInMachine</function>,
            <literal>copy_from_{host,vm}</literal>, and
            <literal>coverage-data</literal>.
```
